### PR TITLE
feat(persistence): add save/load to worker and buttons

### LINKED ISSUE #107: connect save/load button with the actual save/load logic

### PR DESCRIPTION
**Pull request by peipacut**
this PR 1. adds serialisation/deserialisation message types to the worker and 2. gives functionality to the save/restore model buttons. this should also hopefully allow for the automatic save/load task to be started soon.

I also noticed some stuttering in some cases where the state is restored. Hard to repro though and is definitely model related

closes #107 and possibly #71

### CODE REVIEW

File: src/components/ControlBar.tsx [hunk 1]
```diff
@@ -40,23 +43,75 @@ export const RestoreBtn = styled(Button)`
...
-          console.log('Click Restore');
+          fileDialog()
+            .then((files) => {
+              if (files.length > 0) {
+                load(files[0]);
+              }
+            })
+            .catch((e) => {
+              throw e;
+            });
```
Comment (LorenzoValentine) [inline src/components/ControlBar.tsx:114]: The model should be firstly stored in the `local storage`, cause we can not push users to download a file every time when they save the model. Downloading an independent file every time is not a good choice.

**localStorage** *is a web storage object that enables developers to store key-value pairs in a web browser and ensures that this data survives all page refreshes, even when a user closes or restarts a browser*

**Review by LorenzoValentine — CHANGES_REQUESTED** [review]

File: src/components/ControlBar.tsx (hunk 1 above)
Comment (LorenzoValentine) [reply src/components/ControlBar.tsx:114]: https://blog.logrocket.com/localstorage-javascript-complete-guide/

File: src/App.tsx [hunk 2]
```diff
@@ -63,8 +63,13 @@ function App(): React.ReactElement {
       },
     );
     setSimWorker(worker);
+
     const message: IncomingMessage = {
       func: 'init',
+      args: [
+        '/initData/pvf_incomp_44_nonneg/pvf_incomp_44_nonneg_0.json',
```
Comment (Lutra-Fs) [inline src/App.tsx:70]: Should pass model url as well? This one is initial condition

File: src/workers/modelWorkerMessage.ts [hunk 3]
```diff
@@ -7,4 +9,5 @@ export interface OutgoingMessage {
   density?: Float32Array;
   success: boolean;
   matrix?: Float32Array;
+  save: ModelSave;
```
Comment (Lutra-Fs) [inline src/workers/modelWorkerMessage.ts:12]:
```suggestion
	save?: ModelSave;
```

File: package.json [hunk 4]
```diff
@@ -22,6 +22,7 @@
     "@tensorflow/tfjs": "^4.11.0",
     "@vitejs/plugin-react": "^4.0.4",
     "antd": "^5.8.4",
+    "file-dialog": "^0.0.8",
```
Comment (Lutra-Fs) [inline package.json:25]: This dependency is not maintained. DO NOT use this dependency.

File: src/components/ControlBar.tsx [hunk 5]
```diff
@@ -40,23 +43,75 @@ export const RestoreBtn = styled(Button)`
...
+
+  useEffect(() => {
+    modelSaveSubs.push((sav: ModelSave) => {
+      save(sav);
+      console.log('wrote a save');
+    });
+
+    // take the json and have the user download it
+    function save(sav: ModelSave): void {
+      const filename = `${sav.modelType}@${sav.time}.CFSS`;
```
Comment (Lutra-Fs) [inline src/components/ControlBar.tsx:61]: I prefer no custom file extension

File: src/components/ControlBar.tsx [hunk 6]
```diff
@@ -40,23 +43,75 @@ export const RestoreBtn = styled(Button)`
 `;
 
 interface ControlBarProps {
+  modelSaveSubs: Array<(save: ModelSave) => void>;
```
Comment (Lutra-Fs) [inline src/components/ControlBar.tsx:46]: As you are now using a callback to pass the data, it seems like the worker is not needed any more.

File: src/components/ControlBar.tsx [hunk 7]
```diff
@@ -40,23 +43,75 @@ export const RestoreBtn = styled(Button)`
...
-          console.log('Click Save');
+          worker.postMessage({ func: 'serialize' });
         }}
       >
         Save Model
       </SaveBtn>
       <RestoreBtn
         onClick={() => {
-          console.log('Click Restore');
+          fileDialog()
```
Comment (Lutra-Fs) [inline src/components/ControlBar.tsx:106]: Remove this dependency and find some deps maintained.

File: src/components/ControlBar.tsx (hunk 1 above)
Comment (Lutra-Fs) [reply src/components/ControlBar.tsx:114]: Or IndexDB, should probably implement a custom component to allow users select source first (PC/URL/LocalStorage) and create a List for LocalStorage if selected.

File: src/components/ParameterComponents/ParameterLabel.tsx [hunk 8]
```diff
@@ -21,7 +21,7 @@ export default function ParameterLabel(props: {
   tooltip?: string;
 }): React.ReactElement {
   const tooltip: React.ReactElement[] = [];
-  if (props.tooltip) {
+  if (props.tooltip !== null) {
```
Comment (Lutra-Fs) [inline src/components/ParameterComponents/ParameterLabel.tsx:24]: I believe.that the remove of the `!=null` is to resolve ESLint error, does it throw error again?

**Review by Lutra-Fs — CHANGES_REQUESTED** [review]
The subscribe looks promising while the restore needs more work.

File: src/components/ParameterComponents/ParameterLabel.tsx (hunk 8 above)
Comment (peipacut) [reply src/components/ParameterComponents/ParameterLabel.tsx:24]: yeah seems that changing it to `!==null` fixes the issue.

File: src/components/ControlBar.tsx (hunk 1 above)
Comment (peipacut) [reply src/components/ControlBar.tsx:114]: sorry I'm not very familiar with browser IO and just went for the first option I saw. 

Given the size of the model output I think IndexDB might be a better choice, though I'd like if we could implement this in a separate PR.

**Comment by peipacut** [comment]
I've just implemented most of the changes that were pointed out.

As for the file storage options. Could we target that in a separate PR to be completed this week and merge this one?

**Comment by Lutra-Fs** [comment]
> I've just implemented most of the changes that were pointed out.
> 
> 
> 
> As for the file storage options. Could we target that in a separate PR to be completed this week and merge this one?

Can you remove file-dialog first?

**Comment by peipacut** [comment]
> > I've just implemented most of the changes that were pointed out.
> > As for the file storage options. Could we target that in a separate PR to be completed this week and merge this one?
> 
> Can you remove file-dialog first?

sorted

**Review by Lutra-Fs — APPROVED** [review]

**Review by LorenzoValentine — APPROVED** [review]

**Review by Lutra-Fs — APPROVED** [review]

**Pull request merged**